### ISSUE TITLE
[JENKINS-71344] Add retry configuration when fetching from Github

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#3794]
<p>We frequently see flakes using <a href="https://plugins.jenkins.io/git/" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/git/</a> to clone our repositories with error like:</p>

<p><tt>```</tt><br/>
<tt>> git fetch --no-tags --progress git@​github.com:<our_org>/<our_repo>.git +refs/heads/<b>:refs/remotes/origin/</b> # timeout=20</tt><tt>ERROR: Error cloning remote repo 'origin'</tt><tt>hudson.plugins.git.GitException: Command "git fetch --no-tags --progress git@​github.com:<our_org>/<our_repo>.git +refs/heads/<b>:refs/remotes/origin/</b>" returned status code 128:</tt><tt>stdout: Process leaked file descriptors. See {</tt>}<br/>
<tt><a href="https://jenkins.io/redirect/troubleshooting/process-leaked-file-descriptors" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/redirect/troubleshooting/process-leaked-file-descriptors</a></tt><br/>
<tt>for more information</tt><tt>stderr: remote: Enumerating objects: 760803 {</tt>}<br/>
<tt>remote: Enumerating objects: 818003, done.</tt><tt>remote: Compressing objects: 90% (11503/12781) {</tt>}<br/>
{{remote: Compressing objects: 91% (11631/12781) }}<br/>
{{remote: Compressing objects: 92% (11759/12781) }}<br/>
{{remote: Compressing objects: 93% (11887/12781) }}<br/>
{{remote: Compressing objects: 94% (12015/12781) }}<br/>
{{remote: Compressing objects: 95% (12142/12781) }}<br/>
{{remote: Compressing objects: 96% (12270/12781) }}<br/>
{{remote: Compressing objects: 97% (12398/12781) }}<br/>
{{remote: Compressing objects: 98% (12526/12781) }}<br/>
{{remote: Compressing objects: 99% (12654/12781) }}<br/>
{{remote: Compressing objects: 100% (12781/12781) }}<br/>
<tt>remote: Compressing objects: 100% (12781/12781), done.</tt><tt>fatal: Unable to read current working directory: No such file or directory</tt><tt>fatal: index-pack failed</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2042)</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1761)</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$400(CliGitAPIImpl.java:72)</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:442)</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:655)</tt><tt>at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1152)</tt><tt>at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1192)</tt><tt>at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:113)</tt><tt>at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:144)</tt><tt>at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:67)</tt><tt>at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:303)</tt><tt>at hudson.model.ResourceController.execute(ResourceController.java:97)</tt><tt>at hudson.model.Executor.run(Executor.java:429)</tt><tt>ERROR: Error cloning remote repo 'origin'</tt><tt>ERROR: Maximum checkout retry attempts reached, aborting</tt><tt>Finished: FAILURE</tt><br/>
<tt>```</tt></p>

<p>Is it possible to add retry configurations to deal with flakes like this?</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=aaron_wang">aaron_wang</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71344" target="_blank">Add retry configuration when fetching from Github</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: git-plugin
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2023-05-26T12:40:16+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We frequently see flakes using <a href="https://plugins.jenkins.io/git/" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/git/</a> to clone our repositories with error like:</p>

<p><tt>```</tt>
<tt>&gt; git fetch --no-tags --progress git@github.com:&lt;our_org&gt;/&lt;our_repo&gt;.git +refs/heads/<b>:refs/remotes/origin/</b> # timeout=20</tt><tt>ERROR: Error cloning remote repo 'origin'</tt><tt>hudson.plugins.git.GitException: Command "git fetch --no-tags --progress git@github.com:&lt;our_org&gt;/&lt;our_repo&gt;.git +refs/heads/<b>:refs/remotes/origin/</b>" returned status code 128:</tt><tt>stdout: Process leaked file descriptors. See {</tt>}
<tt><a href="https://jenkins.io/redirect/troubleshooting/process-leaked-file-descriptors" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/redirect/troubleshooting/process-leaked-file-descriptors</a></tt>
<tt>for more information</tt><tt>stderr: remote: Enumerating objects: 760803 {</tt>}
<tt>remote: Enumerating objects: 818003, done.</tt><tt>remote: Compressing objects: 90% (11503/12781) {</tt>}
{{remote: Compressing objects: 91% (11631/12781) }}
{{remote: Compressing objects: 92% (11759/12781) }}
{{remote: Compressing objects: 93% (11887/12781) }}
{{remote: Compressing objects: 94% (12015/12781) }}
{{remote: Compressing objects: 95% (12142/12781) }}
{{remote: Compressing objects: 96% (12270/12781) }}
{{remote: Compressing objects: 97% (12398/12781) }}
{{remote: Compressing objects: 98% (12526/12781) }}
{{remote: Compressing objects: 99% (12654/12781) }}
{{remote: Compressing objects: 100% (12781/12781) }}
<tt>remote: Compressing objects: 100% (12781/12781), done.</tt><tt>fatal: Unable to read current working directory: No such file or directory</tt><tt>fatal: index-pack failed</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2042)</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1761)</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$400(CliGitAPIImpl.java:72)</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:442)</tt><tt>at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:655)</tt><tt>at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1152)</tt><tt>at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1192)</tt><tt>at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:113)</tt><tt>at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:144)</tt><tt>at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:67)</tt><tt>at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:303)</tt><tt>at hudson.model.ResourceController.execute(ResourceController.java:97)</tt><tt>at hudson.model.Executor.run(Executor.java:429)</tt><tt>ERROR: Error cloning remote repo 'origin'</tt><tt>ERROR: Maximum checkout retry attempts reached, aborting</tt><tt>Finished: FAILURE</tt>
<tt>```</tt></p>

<p>Is it possible to add retry configurations to deal with flakes like this?</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins version: 2.197.3<br/>
Plugin: <a href="https://plugins.jenkins.io/git/">https://plugins.jenkins.io/git/</a>
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71344?focusedId=438216&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-438216">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Jenkins Pipeline jobs have a retry facility available that you can use, though usually it is a much better choice to resolve the problem at the root of the failure than it is to implement retry.</p>

<p>The log output that you included did not explain why the <tt>git fetch</tt> failed.  Since you have the timeout set to a non-default value, I assume that your repository may be large and need more than the default time to be fetched initially.  You might consider changes that could reduce the time required to fetch the remote repository.  Those changes include:</p>

<ul>
	<li>Use a reference repository on your agents to reduce the amount of data that must be transferred from the remote repository</li>
	<li>Use a narrow refspec in your repository definition to reduce the amount of data that must be transferred from the remote repository</li>
	<li>Honor the configured narrow refspec when fetching the repository</li>
	<li>Use shallow clone to reduce the amount of data that must be transferred from the remote repository</li>
</ul>


<p>Those features are described in the <a href="https://github.com/jenkinsci/git-plugin#clone-extensions" class="external-link" target="_blank" rel="nofollow noopener">"clone extensions"</a> section of the git plugin documentation</p>

<p>They are also described in a <a href="https://youtu.be/jBGFjFc6Jf8" class="external-link" target="_blank" rel="nofollow noopener">10-15 minute recorded talk</a> if you prefer a video version.</p>

<p>You'll need to share the version of the git plugin that you are using and will need to update the version of Jenkins that you are running.  See <a href="https://www.jenkins.io/participate/report-issue/#what-information-to-provide-for-environment-and-description" class="external-link" target="_blank" rel="nofollow noopener">"How to report an issue"</a> for a script that you can use in the script console to report that information.</p>

<p>P.S. My wife loves your product.  She now has 1536 days without missing a day of practicing her Italian on Duolingo (Android).</p>
</pre>
</details>

</ul>
</details>

<p>Jenkins Pipeline jobs have a retry facility available that you can use, though usually it is a much better choice to resolve the problem at the root of the failure than it is to implement retry.</p>

<p>The log output that you included did not explain why the <tt>git fetch</tt> failed.  Since you have the timeout set to a non-default value, I assume that your repository may be large and need more than the default time to be fetched initially.  You might consider changes that could reduce the time required to fetch the remote repository.  Those changes include:</p>

<ul>
	<li>Use a reference repository on your agents to reduce the amount of data that must be transferred from the remote repository</li>
	<li>Use a narrow refspec in your repository definition to reduce the amount of data that must be transferred from the remote repository</li>
	<li>Honor the configured narrow refspec when fetching the repository</li>
	<li>Use shallow clone to reduce the amount of data that must be transferred from the remote repository</li>
</ul>


<p>Those features are described in the <a href="https://github.com/jenkinsci/git-plugin#clone-extensions" class="external-link" target="_blank" rel="nofollow noopener">"clone extensions"</a> section of the git plugin documentation</p>

<p>They are also described in a <a href="https://youtu.be/jBGFjFc6Jf8" class="external-link" target="_blank" rel="nofollow noopener">10-15 minute recorded talk</a> if you prefer a video version.</p>

<p>You'll need to share the version of the git plugin that you are using and will need to update the version of Jenkins that you are running.  See <a href="https://www.jenkins.io/participate/report-issue/#what-information-to-provide-for-environment-and-description" class="external-link" target="_blank" rel="nofollow noopener">"How to report an issue"</a> for a script that you can use in the script console to report that information.</p>

<p>P.S. My wife loves your product.  She now has 1536 days without missing a day of practicing her Italian on Duolingo (Android).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71344?focusedId=438249&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-438249">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>That log output (particularly the "Compressing objects" message from the git server) indicates a large repository and indicates a high risk that you may need to increase the timeout value due to the size of your repository.  A reference repository helps avoid that problem by already placing a base copy of the remote repository on the agent.</p>

<p>The message "fatal: Unable to read current working directory: No such file or directory" may indicate that you're using a network file system for the agent workspace and command line git was unable to get a response from that file system.  It might also indicate that you have some process in the system that is mistakenly reusing the same directory and having the directory be removed while command line git is trying to fetch content into the directory.  Some users have made the mistake of using a fixed directory name for their workspace and then have had multiple jobs attempting to use the same fixed directory name for the workspace.  If you are explicitly declaring the workspace directory yourself, then that could cause "Unable to read current working directory".</p>
</pre>
</details>

</ul>
</details>

<p>That log output (particularly the "Compressing objects" message from the git server) indicates a large repository and indicates a high risk that you may need to increase the timeout value due to the size of your repository.  A reference repository helps avoid that problem by already placing a base copy of the remote repository on the agent.</p>

<p>The message "fatal: Unable to read current working directory: No such file or directory" may indicate that you're using a network file system for the agent workspace and command line git was unable to get a response from that file system.  It might also indicate that you have some process in the system that is mistakenly reusing the same directory and having the directory be removed while command line git is trying to fetch content into the directory.  Some users have made the mistake of using a fixed directory name for their workspace and then have had multiple jobs attempting to use the same fixed directory name for the workspace.  If you are explicitly declaring the workspace directory yourself, then that could cause "Unable to read current working directory".</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71344?focusedId=438261&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-438261">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Retry options are already available from Jenkins Pipeline.  Additional retry options won't be added to the git plugin.</p>
</pre>
</details>

</ul>
</details>

<p>Retry options are already available from Jenkins Pipeline.  Additional retry options won't be added to the git plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aaron_wang:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71344?focusedId=438264&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-438264">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<ul>
	<li>Based on my search on <a href="https://stackoverflow.com/questions/50758221/jenkins-git-fetch-intermittent-failure" class="external-link" target="_blank" rel="nofollow noopener">Stackoverflow</a>, I think this happens when the git server's `MaxStartups` sshd_config is set to be more strict than the number of `git fetch` we are running in parallel. We don't run our own git server and use vanilla Github private repos, and as far as I can tell, I don't see a way to change the sshd_config on the Github side. It also happens that the `git fetch` fails with different error messages every time, I just took one example here, so I think having some client-side retries will really help to work around the problem.</li>
	<li>By Jenkin's retry, are you referring to <a href="https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#retry-retry-the-body-up-to-n-times?" class="external-link" target="_blank" rel="nofollow noopener">https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#retry-retry-the-body-up-to-n-times?</a> The problem is that our Jenkins pipeline groovy script is read from a github repo using this Git Jenkins plugin. Intermittent `git fetch` failures like this often happen during the initial clone, by which time the groovy pipeline script isn't yet loaded and we cannot use Jenkin's pipeline step retry as a result. So having native retry support from this plugin will be very helpful!</li>
	<li>Thanks for the tips on faster clone. We already use shallow clone with `Fetch tags` unselected. This particular pipeline can't use a narrower refspec because it's also running at pre-merge time, and we don't know the branch name ahead of time. That said, we do notice having a lot of remote branches causes us delays when cloning and we are working to delete some of our old remote branches.</li>
	<li>Updated original post to include full error log. (I'm working with our Jenkins admin to get info from <a href="https://jenkins-ci.duolingo.com/systemInfo" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins-ci.duolingo.com/systemInfo</a> since I don't have permission to do so)</li>
</ul>

</pre>
</details>

</ul>
</details>

<ul>
	<li>Based on my search on <a href="https://stackoverflow.com/questions/50758221/jenkins-git-fetch-intermittent-failure" class="external-link" target="_blank" rel="nofollow noopener">Stackoverflow</a>, I think this happens when the git server's `MaxStartups` sshd_config is set to be more strict than the number of `git fetch` we are running in parallel. We don't run our own git server and use vanilla Github private repos, and as far as I can tell, I don't see a way to change the sshd_config on the Github side. It also happens that the `git fetch` fails with different error messages every time, I just took one example here, so I think having some client-side retries will really help to work around the problem.</li>
	<li>By Jenkin's retry, are you referring to <a href="https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#retry-retry-the-body-up-to-n-times?" class="external-link" target="_blank" rel="nofollow noopener">https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#retry-retry-the-body-up-to-n-times?</a> The problem is that our Jenkins pipeline groovy script is read from a github repo using this Git Jenkins plugin. Intermittent `git fetch` failures like this often happen during the initial clone, by which time the groovy pipeline script isn't yet loaded and we cannot use Jenkin's pipeline step retry as a result. So having native retry support from this plugin will be very helpful!</li>
	<li>Thanks for the tips on faster clone. We already use shallow clone with `Fetch tags` unselected. This particular pipeline can't use a narrower refspec because it's also running at pre-merge time, and we don't know the branch name ahead of time. That said, we do notice having a lot of remote branches causes us delays when cloning and we are working to delete some of our old remote branches.</li>
	<li>Updated original post to include full error log. (I'm working with our Jenkins admin to get info from <a href="https://jenkins-ci.duolingo.com/systemInfo" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins-ci.duolingo.com/systemInfo</a> since I don't have permission to do so)</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-71344?focusedId=438265&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-438265">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>I think this happens when the git server's `MaxStartups` sshd_config is set to be more strict than the number of `git fetch` we are running in parallel. We don't run our own git server and use vanilla Github private repos, and as far as I can tell, I don't see a way to change the sshd_config on the Github side. </p></blockquote>

<p>I think you will be much better served by checking your own network and system configurations rather than assuming that GitHub has a configuration error.  GitHub is the organization that has found ways to make millions of git ssh connections reliable to a single hostname, github.com.  I would be truly astonished if they had a configuration error in their ssh configuration.  While it is possible, it seems unlikely</p>

<p>When command line git reports <tt>Unable to read current working directory</tt>, as far as I understand it, that is a local failure, not a failure on the remote git server.  If a directory no longer exists on the local server, that seems like it is best to investigate locally.</p>

<p>I'm not willing to implement retry in the git plugin.  It is exceptionally difficult to implement it correctly, exceptionally difficult to test, and exceptionally difficult to maintain once implemented.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>I think this happens when the git server's `MaxStartups` sshd_config is set to be more strict than the number of `git fetch` we are running in parallel. We don't run our own git server and use vanilla Github private repos, and as far as I can tell, I don't see a way to change the sshd_config on the Github side. </p></blockquote>

<p>I think you will be much better served by checking your own network and system configurations rather than assuming that GitHub has a configuration error.  GitHub is the organization that has found ways to make millions of git ssh connections reliable to a single hostname, github.com.  I would be truly astonished if they had a configuration error in their ssh configuration.  While it is possible, it seems unlikely</p>

<p>When command line git reports <tt>Unable to read current working directory</tt>, as far as I understand it, that is a local failure, not a failure on the remote git server.  If a directory no longer exists on the local server, that seems like it is best to investigate locally.</p>

<p>I'm not willing to implement retry in the git plugin.  It is exceptionally difficult to implement it correctly, exceptionally difficult to test, and exceptionally difficult to maintain once implemented.</p>
